### PR TITLE
[SYCL] Move internal headers to source dir

### PR DESCRIPTION
**Pull request by alexbatashev**
Move internal headers from include/CL/sycl to source directory to prevent implementation details leak to user application and enforce stable ABI.

A few more changes were applied to make the movement possible:
- `addHostAccessorAndWait` functions in accessor to avoid calls to RT internals from header file
- Move `getImageInfo` function to cpp
- Move buffer size acquisition from buffer constructor to SYCLMemObjT cpp to avoid calls to PI
- `getPluginFromContext` function in context
- Standard containers replaced with SYCL variants in sycl_mem_obj_i.hpp. Unique ptr replaced with shared
- A few implementations moved from queue.hpp to queue.cpp
- Some LIT tests temporarily include implementaion specific headers. They will be converted to unit tests later.
- clang-format applied

Signed-off-by: Alexander Batashev <alexander.batashev@intel.com>

### CODE REVIEW

File: sycl/source/detail/stream_impl.cpp [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 #include <CL/sycl/detail/stream_impl.hpp>
 #include <cstdio>
+#include <detail/scheduler/scheduler.hpp>
```
Comment (bader) [inline sycl/source/detail/stream_impl.cpp]: According to the coding style guidelines system header includes should go last. https://llvm.org/docs/CodingStandards.html#include-style
```c++
#include <CL/sycl/detail/stream_impl.hpp>
#include <detail/scheduler/scheduler.hpp>

#include <cstdio>
```

File: sycl/source/event.cpp [hunk 2]
```diff
@@ -8,11 +8,11 @@
...
 #include <CL/sycl/context.hpp>
 #include <CL/sycl/detail/common.hpp>
-#include <CL/sycl/detail/event_impl.hpp>
 #include <CL/sycl/detail/pi.hpp>
-#include <CL/sycl/detail/scheduler/scheduler.hpp>
-#include <CL/sycl/info/info_desc.hpp>
 #include <CL/sycl/event.hpp>
+#include <CL/sycl/info/info_desc.hpp>
+#include <detail/event_impl.hpp>
+#include <detail/scheduler/scheduler.hpp>
```
Comment (bader) [inline sycl/source/event.cpp]: Please, remove this empty line and re-apply clang-format.

File: sycl/test/lit.cfg.py [hunk 3]
```diff
@@ -68,7 +68,7 @@
     config.environment['OCL_ICD_FILENAMES'] = os.environ['OCL_ICD_FILENAMES']
 
 config.substitutions.append( ('%clang_cc1', ' ' + config.clang + ' -cc1 ') )
-config.substitutions.append( ('%clangxx', ' ' + config.clangxx + ' -I'+config.opencl_include ) )
+config.substitutions.append( ('%clangxx', ' ' + config.clangxx + ' -I'+config.opencl_include + ' -I'+config.sycl_source_dir ) )
```
Comment (bader) [inline sycl/test/lit.cfg.py]: I think all clang variables must have the same set of include search paths.
Please, apply to `%clang_cl` and `%clang`.
I also think we should not use `%clang_cc1` in SYCL runtime library tests, so it's better to remove this variable from sycl lit config file.

File: sycl/unittests/CMakeLists.txt [hunk 4]
```diff
@@ -22,7 +22,10 @@ function(add_sycl_unittest test_dirname)
       LLVMTestingSupport
       OpenCL-Headers
     )
-  target_include_directories(${test_dirname} PRIVATE SYSTEM ${sycl_inc_dir})
+  target_include_directories(${test_dirname} PRIVATE SYSTEM
+      ${sycl_inc_dir}
+      ${CMAKE_SOURCE_DIR}/../sycl/source/
```
Comment (bader) [inline sycl/unittests/CMakeLists.txt]: Please, define `SYCL_SOURCE_DIR` variable in root cmake file to re-use it here.

File: sycl/test/lit.cfg.py (hunk 3 above)
Comment (alexbatashev) [reply sycl/test/lit.cfg.py]: @bader I'm going to rewrite tests that use something from source dir as unit tests. Then I'll remove this flag. Personally, I'm not ok with end-to-end tests accessing some internals. Thus, I see no point in adding aforementioned include path to other commands. `%clang_cc1` will also be removed in a separate patch. Are you ok with that?

File: sycl/source/context.cpp [hunk 5]
```diff
@@ -8,14 +8,14 @@
...
 #include <CL/sycl/detail/common.hpp>
-#include <CL/sycl/detail/context_impl.hpp>
 #include <CL/sycl/device.hpp>
 #include <CL/sycl/device_selector.hpp>
 #include <CL/sycl/exception.hpp>
 #include <CL/sycl/exception_list.hpp>
 #include <CL/sycl/platform.hpp>
 #include <CL/sycl/stl.hpp>
 #include <algorithm>
+#include <detail/context_impl.hpp>
```
Comment (bader) [inline sycl/source/context.cpp]: Apply following pattern:
```c++
#include <CL/sycl/...> // SYCL API headers 
// empty line no necessary as lexicographical order will these includes in the right order - up to you
#include <detail/...> // SYCL internal headers located in source directory
// empty line
#include <alrogithm> // C++ STL headers
```

File: sycl/unittests/thread_safety/ThreadUtils.h [hunk 6]
```diff
@@ -1,3 +1,4 @@
+#include <detail/scheduler/scheduler.hpp>
```
Comment (bader) [inline sycl/unittests/thread_safety/ThreadUtils.h]:
```suggestion
#include <detail/scheduler/scheduler.hpp>

```

File: sycl/test/lit.cfg.py (hunk 3 above)
Comment (bader) [reply sycl/test/lit.cfg.py]: Are you going to rewrite the tests in the scope of this PR or separately?

Comment (alexbatashev) [reply sycl/test/lit.cfg.py]: Separately

**Comment by bader** [comment]
Please, resolve merge conflicts.

**Comment by alexbatashev** [comment]
> Please, resolve merge conflicts.

Done.

File: sycl/include/CL/sycl/accessor.hpp [hunk 7]
```diff
@@ -16,7 +16,6 @@
 #include <CL/sycl/detail/generic_type_traits.hpp>
 #include <CL/sycl/detail/image_accessor_util.hpp>
 #include <CL/sycl/detail/image_ocl_types.hpp>
-#include <CL/sycl/detail/queue_impl.hpp>
```
[the remark this answers is not in the document]
Comment (alexbatashev) [reply sycl/include/CL/sycl/accessor.hpp]: @romanovvlad yes

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 8]
```diff
@@ -259,6 +256,8 @@ class SYCLMemObjT : public SYCLMemObjI {
     MAllocator->setAlignment(RequiredAlign);
   }
 
+  static size_t getBufSizeForContext(ContextImplPtr Context, cl_mem MemObject);
```
[the remark this answers is not in the document]
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: Done

File: sycl/include/CL/sycl/accessor.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/CL/sycl/accessor.hpp]: Let's move at least AccessorImplHost to the private headers.

[the remark this answers is not in the document]
Comment (alexbatashev) [reply sycl/include/CL/sycl/accessor.hpp]: @romanovvlad I'll do it as a separate patch.

**Comment by bader** [comment]
@alexbatashev, please, fix failed/pending jobs.

**Comment by alexbatashev** [comment]
> @alexbatashev, please, fix failed/pending jobs.

These are assert and opencl_interop. I restarted tests, hopefully, they won't fail me next time.

**Pull request merged**